torch/_dynamo/variables/torch.py

```diff
@@ -1046,29 +952,38 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
             graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
 
             def speculate_branch(branch):
+                # NB: 0 is predicate
+                ix = 1 if branch else 2
                 try:
-                    # NB: 0 is predicate
-                    ix = 1 if branch else 2
-                    return old_speculate_subgraph(
-                        args[ix], operands, graph_checkpoint, checkpoint
+                    ret_val, ret_graph, ret_lifted_freevars = speculate_subgraph(
+                        tx, args[ix], operands, graph_checkpoint, checkpoint
+                    )
+                except Exception as e:
+                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e)) from e
+                if not isinstance(ret_val, TensorVariable):
+                    raise UserError(
+                        UserErrorType.DYNAMIC_CONTROL_FLOW,
+                        "Expected branch out type to be a single tensor but got {}".format(
+                            str(ret_val.python_type())
+                        ),
                     )
-                except ArgsMismatchError as e:
-                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e))
+                return ret_val, ret_graph, ret_lifted_freevars
 
-            (
-                true_r,
-                true_graph,
-                true_guards,
-                true_nn_modules_context,
-                true_cmp,
-            ) = speculate_branch(True)
-            (
-                false_r,
-                false_graph,
-                false_guards,
-                false_nn_modules_context,
-                false_cmp,
-            ) = speculate_branch(False)
+            (true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
+
+            tx.output.side_effects.prune_dead_object_new(tx)
+            state_after_true_branch = tx.copy_graphstate()
+            true_guards = state_after_true_branch.output.guards
+            true_nn_modules = state_after_true_branch.output.nn_modules
+            true_cmp = get_comparable_state(state_after_true_branch)
```

Comment:
Is the `true_cmp` object actually used for anything? Ditto for `false_cmp`.

Reply:
Nah just trying to preserve the old code.

Reply:
They are used to add tracking tensors to global tensor tracking list. So it seems needed?

Reply:
Where does that logic happen?

Reply:
https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/torch.py#L1138